core/chains/solana/soltxm/pendingtx.go

```diff
@@ -0,0 +1,166 @@
+package soltxm
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+	"golang.org/x/exp/maps"
+)
+
+type PendingTxContext interface {
+	Add(sig solana.Signature, cancel context.CancelFunc) error
+	Remove(sig solana.Signature)
+	ListAll() []solana.Signature
+	Expired(sig solana.Signature, lifespan time.Duration) bool
+	// state change hooks
+	OnSuccess(sig solana.Signature)
+	OnError(sig solana.Signature, errType int) // match err type using enum
+}
+
+var _ PendingTxContext = &pendingTxContext{}
+
+type pendingTxContext struct {
+	cancelBy  map[solana.Signature]context.CancelFunc
+	timestamp map[solana.Signature]time.Time
+	lock      sync.RWMutex
+}
+
+func newPendingTxContext() *pendingTxContext {
+	return &pendingTxContext{
+		cancelBy:  map[solana.Signature]context.CancelFunc{},
+		timestamp: map[solana.Signature]time.Time{},
+	}
+}
+
+func (c *pendingTxContext) Add(sig solana.Signature, cancel context.CancelFunc) error {
+	// already exists
+	c.lock.RLock()
+	if c.cancelBy[sig] != nil {
+		c.lock.RUnlock()
+		return errors.New("signature already exists")
+	}
+	c.lock.RUnlock()
+
+	// save cancel func
+	c.lock.Lock()
```

Comment:
Could another goroutine grab the lock in between and put their own cancel func (which would be overwritten)? Generally there is a defensive double-check after upgrading to the write lock:
```suggestion
	c.lock.RUnlock()

	// save cancel func
	c.lock.Lock()
	if c.cancelBy[sig] != nil {
		c.lock.Unlock()
		return errors.New("signature already exists")
	}
```